src/v/bytes/iobuf.h

```diff
@@ -40,15 +40,56 @@
  *
  * General sharing-mutation caveat:
  *
+ * iobuf has more complicated mutation and cross-shard sharing rules, compared
+ * to most other types such as int or std::string. The underlying cause in both
+ * cases is that two different iobuf objects may share one or more underlying
+ * buffers, and hence operations on one iobuf may be visible to the other.
+ *
  * Operations such as share(), copy() and appending an iobuf or other compatible
  * buffer type to an iobuf may be zero-copy, in the sense that some or all of
  * the payload bytes may be shared between multiple iobufs (or between an iobuf
  * and a compatible buffer type like ss::temporary_buffer<>). The sharing occurs
  * at the fragment level.
  *
- * Be careful when any zero-copy operations are used as iobuf
- * does not perform copy-on-write, Therefore changes will be visible to all
- * iobufs that share the backing fragments.
+ * We say that two or more iobuf objects which share fragments have "internal
+ * sharing" and between such iobufs the following restrictions apply:
+ *
+ * BYTE MUTATION CAVEAT
+ *
+ * You should not write into the bytes held by an iobuf if it is internally
+ * shared with another buffer, since the updates will potentially been seen
+ * by both iobufs.
+ *
+ * On the other hand, two iobufs that have internal sharing will behave
+ * independently with respect to "structural updates", which are all mutations
+ * except for writing into the buffer itself. For example, if one iobuf is
+ * created as a copy of another via share() method, they will have full internal
+ * sharing, but appending to one buffer will not been seen Be careful when any
+ * zero-copy operations are used as iobuf does not perform copy-on-write,
+ * Therefore changes will be visible to all iobufs that share the backing
+ * fragments.
+ *
+ * CROSS-SHARD SHARING CAVEAT
+ *
+ * Two iobufs which have internal sharing should not be accessed concurrently on
+ * different shards. Note that this is a much stronger condition than the usual
+ * thread-safety requirements for C++ objects since this applies to different
+ * objects with (potentially hidden) internal sharing, while the usual rules
+ * apply only to sharing of the _same_ object.
+ *
+ * More formally and slightly stricter than the above: every iobuf has an
+ * "origin" shard which cannot be changed and it must only be accessed on that
+ * shard: access from an other shard is an error which may or may not be
+ * detected. An iobuf's origin shard is set at construction, as documented
+ * in the method doc (for example, the default constructor sets the origin
+ * shard to the current one, while the move constructor inherits the origin
+ * shard from the source and so on).
+ *
+ * The only safe way to get the contents of an iobuf from one shard to another
+ * is to pass the iobuf to the other shard and then call copy() on it, which is
```

Comment:
This doesn't seem like a safe interface to me? 

It means the `submit_to` caller always has to wait until the `submit_to` future resolves to do anything with its iobuf and hope the `submit_to` side is not keeping the original iobuf around somehow? It seems as unsafe to me as the more idiomatic way of just doing copy/share first and then moving the result to the other shard. It further pessimizes the cases where no copy is needed and the iobuf is simply EOL on the source shard.

I think the oncore stuff is just broken and doesn't really work as is. I tried to [extend](https://github.com/redpanda-data/redpanda/pull/14024) it in the past as well.

This really needs language support, maybe there is some thread_local hackery that might work?

Reply:
> It means the submit_to caller always has to wait until the submit_to future resolves to do anything with its iobuf and hope the submit_to side is not keeping the original iobuf around somehow?

Right, though waiting until the future resolves is almost always the existing use case for submit_to and related calls? 

That said, I wrote this as a codification of the oncore existing restrictions of the class, and the existing behavior of the share/copy/move ctor methods: the thing that creates the new iobuf (here, `copy()` needs to happen on the target shard since it grabs its origin shard ID implicitly from the current shard.

As far as I can tell, iobuf does not currently allow _any_ reasonable way to move an iobuf from one shard to another: the origin ID is preserved on move, so you can never get a moved to iobuf on another shard with the right origin ID. It may only happen to work in some cases because the verify check wasn't uniformly added everywhere so if you avoid the methods where it does exist you at least won't get an assert (but you may crash).

I don't think it's a slam dunk though that `copy()`-on-source is definitely better at least if we have any concept of origin shard: consider the case where you do an `invoke_on` (all shards), you really want to create N copies (at east with the current restriction), one for each shard. In that case copy() on the target just works: makes the right number of copies of the iobuf with right shard ID.

> It further pessimizes the cases where no copy is needed and the iobuf is simply EOL on the source shard.

Yes, but as above this is "already the case" and don't think there's any way to move safely currently.

> I think the oncore stuff is just broken and doesn't really work as is.

Well the restriction is "too strict", but we already have it so I'm trying to lean on that for safety at the cost of re-enforcing the existing pessimization. So it is at least partly workable in the sense that it has been like this since iobuf introduction.

That said, totally open to better approaches here too, either for the doc or the implementation. Just relying on 100% adherence to complex and mostly undocumented lifetime runs isn't going to cut it, I think: especially because the one of the main failure mode occurs due to a non-atomic increment race which is going to be very hard to pick up with testing and likely to slip through to production.

> I tried to https://github.com/redpanda-data/redpanda/pull/14024 it in the past as well.

Ah, I wish I had seen that. Yeah it's the same vein. There's also more discussion in https://redpandadata.atlassian.net/browse/CORE-7061 if you hadn't seen it.

> This really needs language support, maybe there is some thread_local hackery that might work?

Sure, we should be open to anything here. Maybe like @ballard26 said it's easier just to make the refcounting atomic, which would make the semantics more sane and hence easier to document :). That does look like a potentially invasive change performance-wise though since I think it means temporary_buffer needs use atomic refcounting everywhere in seastar (i.e., not opt-in e.g., via template parameter) since we buffers created and consumed in seastar to be like that from the start: plus the work of combing through all the `deleter` implementations to see if they are truly compatible with cross-shard use beyond the refcounting, since they can do arbitrary things at deletion time.

Reply:
> As far as I can tell, iobuf does not currently allow any reasonable way to move an iobuf from one shard to another: the origin ID is preserved on move, so you can never get a moved to iobuf on another shard with the right origin ID. It may only happen to work in some cases because the verify check wasn't uniformly added everywhere so if you avoid the methods where it does exist you at least won't get an assert (but you may crash).

> Yes, but as above this is "already the case" and don't think there's any way to move safely currently.

Right yes, I guess I should have clarified that I was thinking more about the release build usecase or with the oncore check removed as I think it doesn't make much sense in its current form as described otherwise. I do very much suspect it wasn't added to all methods because it would break some common current usage patterns that are fine from a thread safety perspective.

> I don't think it's a slam dunk though that copy()-on-source is definitely better at least if we have any concept of origin shard: consider the case where you do an invoke_on (all shards), you really want to create N copies (at east with the current restriction), one for each shard. In that case copy() on the target just works: makes the right number of copies of the iobuf with right shard ID.

In that example, is a copy even needed at all? If it's safe to copy/read from N other threads in parallel then it would also be safe to just read directly without copying for the duration of the `invoke_on_all` call? It would require const == threadsafe, not sure whether that is given at least today.

> said it's easier just to make the refcounting atomic ... potentially invasive change performance-wise though

I do think that is the best way forward. Back when I was looking into this because of the same bug/misuse in a different part of the code this was my conclusion as well. Note I think this could be a performance gain as it would mean that we can drop the extra copies on the produce and fetch path which are likely a lot more expensive. 

For the checks, I am wondering whether we can add a method along the signature of:

```
iobuf&& move_to_shard(shard_id target_shard_id)
```

That invalidates the current object and creates a new one with the source shard id adapted. Then we could possibly add the checks in more places (outside of the constructors which are probably still needed implicitly as seastar might move the lambda under the hood). Haven't really thought this through.

Reply:
But no, `move_to_shard` is never really safe because at least in its current form the iobuf doesn't know whether it uniquely owns the temporary_buffer (in https://github.com/redpanda-data/redpanda/issues/13447#issuecomment-1752007143 the issue was definitely that we could have already gotten a `shared` buffer out of `ss::input_stream::read_up_to`).

So I guess this really already pessimizes this case:

> It further pessimizes the cases where no copy is needed and the iobuf is simply EOL on the source shard.

as it can never be correct unfortunately. And in that sense `share` with oncore check failing anywhere is a bug.

So possibly `copy_to_shard(shard_id target_shard_id)` might be a good interface but in practice just doing the `copy` in the `submit_to` callback is just easier then.

Reply:
> Right yes, I guess I should have clarified that I was thinking more about the release build usecase or with the oncore check removed as I think it doesn't make much sense in its current form as described otherwise.

Well I feel like we shouldn't do be doing anything in the release build that would assert in the debug build, right? Both "in principle" but also actually it should be unlikely since we run (some subset of) our tests in debug and those don't assert. So I would say that the methods that have the oncore check are off-limits for cross-origin calls regardless of the build mode.

That aside, I basically agree with you that the oncore check seems like it's not really doing the right thing or at least pessimizes things unnecessarily. So I'll consider this on hold for now.

Reply:
> as it can never be correct unfortunately.

Right. Though some correct cases can be detected, conservatively, though it would require more tracking. E.g., every operation either potentially causes shared buffers to be added to the iobuf, or not. E.g., the constructors that just pass char*/size which is copied into the iobuf don't result in sharing, and something like the move-ctor should inherit the sharing state. `clear()` removes all sharing, etc. So you track this boolean state "not shared vs maybe shared" and only assert if it's in the maybe shared state or some other kind of logic. E.g., moving across shards is safe for not shared stuff.

> As are the failures in this PR?

I did look at two cases and they both looked safe in the sense above: on one shard an iobuf is created "unshared" (e.g., as the result of parsing some json), then it is moved to another shard (produce flow in this case).